Tests/test_map.gd

```diff
@@ -12,3 +12,125 @@ func test_map_gen() -> void:
 	for couple: Array[int] in expected_exists_array:
 		var _room: RoomBase = test_map.rooms[couple[0]][couple[1]]
 		assert_not_null(_room, "Expected not null at %s but got %s" % [couple, _room])
+
+## Rule 0: Rooms should not have no event
+func test_rule_0_map_gen() -> void:
+	var expected_exists_array: Array[Array] = [[0,2],[1,1],[1,2],[1,3],[2,0],[2,1],[2,2],[2,3],[2,4],[3,1],[3,2],[3,3],[4,2]]
+
+	for couple: Array[int] in expected_exists_array:
+		var _room_event: EventBase = test_map.rooms[couple[0]][couple[1]].room_event
+		assert_not_null(_room_event, "Expected not null event at %s but got %s" % [couple, _room_event])
+
+## Rule 1: No Heal room before half of the map
+func test_rule_1_map_gen() -> void:
+	var current_room: RoomBase
+	for index_height: int in range(int(test_map.rooms.size()/2.0)):
+		for index_width: int in range(test_map.rooms[index_height].size()):
+			current_room = test_map.rooms[index_height][index_width]
+			
+			if current_room == null:
+				continue
+			
+			assert_not_same("heal", current_room.room_event.get_event_name(), 
+				"Should not have a Heal room before half of the map but found %s at %s" % [current_room.room_event.get_event_name(), current_room.room_position])
+				
+## Rule 2: No consecutive shop or heal rooms
+func test_rule_2_map_gen() -> void:
+	var no_consecutive_room_event: Array[String] = ["shop", "heal"]
+	var current_room: RoomBase
+	
+	# Don't care about last floor
+	for index_height: int in range(test_map.rooms.size() - 1):
+		for index_width: int in range(test_map.rooms[index_height].size()):
+			current_room = test_map.rooms[index_height][index_width]
+			
+			# If current room is not a Heal or Shop room, ignore
+			if current_room == null or (not current_room.room_event.get_event_name() in no_consecutive_room_event):
+				continue
+			print(index_height, " ", index_width)
+			
+			# Get child room type and compare
+			for delta_x: int in [-1, 0, 1]:
+				var child_x: int = current_room.room_position.x + delta_x
+				var child_y: int = current_room.room_position.y + 1
+				if child_y >= 0 && child_y < test_map.rooms.size() && child_x >= 0 && child_x < test_map.rooms[child_y].size() && test_map.rooms[child_y][child_x] != null:
+					var consecutive: bool = (current_room.room_event.get_event_name() in no_consecutive_room_event && 
+											 current_room.room_event.get_event_name() == test_map.rooms[child_y][child_x].room_event.get_event_name())
+			
+					assert_false(consecutive, "Should not have two consecutive Heal rooms or two consecutive Shop rooms but found %s at %s and %s at %s" % 
+								 [current_room.room_event.get_event_name(), current_room.room_position, test_map.rooms[child_y][child_x].room_event.get_event_name(), test_map.rooms[child_y][child_x].room_position])
+
+
+## Rule 3: There must be at least 2 room types among destinations of Rooms that have 2 or more Paths going out.
```

Comment:
According to what you wrote in the rule checking function of map generation, we also make sure the two leftmost and rightmost events are different, but here you only say that it counts only for rooms that have 2 or more paths going out, which one is it ?